arch/x86_64/mcount-insn.c

```diff
@@ -31,66 +31,102 @@ void mcount_disasm_finish(struct mcount_disasm_engine *disasm)
 	cs_close(&disasm->engine);
 }
 
+#define INSTRUMENT_FAIL_UNKNOWN		0x00000001L
+#define INSTRUMENT_FAIL_NOOPRND		0x00000001L << 1
+#define INSTRUMENT_FAIL_RELJMP		0x00000001L << 2
+#define INSTRUMENT_FAIL_RELCALL		0x00000001L << 3
+#define INSTRUMENT_FAIL_PICCODE		0x00000001L << 4
+
+#define OP_GROUP_JMP			1
+#define OP_GROUP_CALL			2
+
+void print_instrument_fail_msg(int reason)
+{
+	if (reason & INSTRUMENT_FAIL_NOOPRND) {
+		pr_dbg3("\tNot supported opcode without operand\n");
+	}
+	if (reason & INSTRUMENT_FAIL_RELJMP) {
+		pr_dbg3("Not supported opcode that jump to relative address\n");
+	}
+	if (reason & INSTRUMENT_FAIL_RELCALL) {
+		pr_dbg3("Not supported opcode that call to relative address\n");
+	}
+	if (reason & INSTRUMENT_FAIL_PICCODE) {
+		pr_dbg3("Not supported Position Independent Code\n");
+	}
+}
+
 /*
  * check whether the instruction can be executed regardless of its location.
  * returns false when instructions are not suitable for dynamic patch.
  *
  * TODO: this function is incomplete and need more classification.
  */
-static bool check_instrumentable(struct mcount_disasm_engine *disasm,
+static int check_instrumentable(struct mcount_disasm_engine *disasm,
 				 cs_insn *insn)
 {
 	int i;
 	cs_x86 *x86;
 	cs_detail *detail;
-	bool jmp_or_call = false;
-	bool status = false;
+	int jmp_or_call = 0;
+	int status = 0;
 
 	/*
 	 * 'detail' can be NULL on "data" instruction
 	 * if SKIPDATA option is turned ON
 	 */
-	if (insn->detail == NULL)
-		return false;
+	if (insn->detail == NULL) {
+		status = INSTRUMENT_FAIL_UNKNOWN;
+		goto out;
+	}
 
 	detail = insn->detail;
 
 	for (i = 0; i < detail->groups_count; i++) {
-		if (detail->groups[i] == CS_GRP_CALL ||
-		    detail->groups[i] == CS_GRP_JUMP) {
-			jmp_or_call = true;
-		}
+		if (detail->groups[i] == CS_GRP_CALL)
+			jmp_or_call = OP_GROUP_CALL;
+		else if (detail->groups[i] == CS_GRP_JUMP)
+			jmp_or_call = OP_GROUP_JMP;
 	}
 
 	x86 = &insn->detail->x86;
 
-	/* no operand: disallow just to be safer for now */
 	if (!x86->op_count)
-		return true;
+		goto out;
 
 	for (i = 0; i < x86->op_count; i++) {
 		cs_x86_op *op = &x86->operands[i];
 
 		switch((int)op->type) {
-		case X86_OP_REG:
-			status = true;
-			break;
-		case X86_OP_IMM:
-			if (jmp_or_call)
-				return false;
-			status = true;
-			break;
-		case X86_OP_MEM:
-			if (op->mem.base == X86_REG_RIP ||
-			    op->mem.index == X86_REG_RIP)
-				return false;
-
-			status = true;
-			break;
-		default:
-			break;
+			case X86_OP_REG:
+				continue;
+
+			case X86_OP_IMM:
+				if (jmp_or_call > 0) {
+					if (jmp_or_call == OP_GROUP_CALL)
+						status |= INSTRUMENT_FAIL_RELCALL;
+					else
+						status |= INSTRUMENT_FAIL_RELJMP;
+					goto out;
+				}
+				continue;
```

Comment:
Please consider reducing indent levels by bailing out the short case.

```c
case X86_OP_IMM:
        if (!jmp_or_call)
                continue;
        ...
```

Reply:
thanks for advise